File: src/schatd/connection.cpp
/* $Id$
 * IMPOMEZIA Simple Chat
 * Copyright © 2008-2009 IMPOMEZIA <schat@impomezia.com>
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#include <QtCore>
#include <boost/bind.hpp>

#define SCHAT_DEBUG

#include "simpleclienthandler.h"
#include "asio/asio.hpp"
#include "connection.h"
#include "protocol4.h"

/*!
 * \brief Приватный D-класс для класса Connection.
 */
class Connection::Private
{
public:
  Private(Connection *parent, asio::io_service &ioService)
  : handler(0),
  checkTimer(0),
  pingTimer(ioService),
  socket(ioService),
  q(parent),
  pingState(Connection::WaitPing),
  bodySize(0),
  opcode(0),
  m_state(WaitGreeting)
  {}

  ~Private()
  {
    if (handler) delete handler;
    if (checkTimer) delete checkTimer;
  }

  inline State state() const        { return m_state; }
  inline void setState(State state) { m_state = state; }
  void check(asio::error_code &err);
  void close();
  void readBody(const asio::error_code &err, int bytes);
  void readHeader(const asio::error_code &err, int bytes);
  void send();
  void start();

  AbstractProtocolHandler *handler;                ///< Обработчик протокола.
  asio::deadline_timer *checkTimer;                ///< Таймер, обслуживающий соединение.
  asio::deadline_timer pingTimer;                  ///< Таймер, для пинг-запросов для состояния соединения.
  asio::ip::tcp::socket socket;                    ///< Сокет обслуживающий данное соединение.
  char bodyBuffer[protocol::packet::MaxSize];      ///< Буфер для чтения тела пакета.
  char headerBuffer[protocol::packet::HeaderSize]; ///< Буфер для заголовка пакета.
  char sendBuffer[protocol::packet::MaxSize];      ///< Буфер отправки пакета.
  Connection *q;                                   ///< Указатель на родительский объект.
  PingState pingState;                             ///< Текущее состояние механизма проверки соединения.
  QQueue<QByteArray> sendQueue;                    ///< Очередь пакетов для отправки.
  quint16 bodySize;                                ///< Размер тела пакета.
  quint16 opcode;                                  ///< Опкод пакета.

private:
  bool detect();

  State m_state;                                   ///< Статус соединения.
};


/*!
 * Обнаружение мёртвых соединений без инициированного рукопожатия.
 * Функция вызывается спустя \b 20 секунд.
 *
 * Если рукопожатие не было инициировано соединение разрывается.
 * В противном случае запускается пинг-таймер.
 *
 * \param err Код ошибки таймера.
 */
void Connection::Private::check(asio::error_code &err)
{
  DEBUG_OUT("Connection::Private::check()" << this)
//  qDebug() << "[2]" << QDateTime::currentDateTime().toString("mm:ss") << state();

  if (!err) {
//    if (d->state() == WaitGreeting) {
//      handleClose();
//      return;
//    }

//    if (d->state() != WaitClose) {
//      d->timer.expires_from_now(boost::posix_time::millisec(50));
//      d->timer.async_wait(boost::bind(&Connection::checkGreeting, shared_from_this(), _1));
//    }
//      startPing(WaitPing, schat::PingInterval);
  }

  if (checkTimer) {
    delete checkTimer;
    checkTimer = 0;
  }
}


/*!
 * Закрытие соединения.
 * Состояние устанавливается в Connection::WaitClose, останавливается
 * таймер и останавливается передача и приём данных и сокет закрывается.
 *
 * \todo Необходимо уведомлять ядро о причине разъединения.
 */
void Connection::Private::close()
{
  if (state() == WaitClose)
    return;

  setState(WaitClose);
  if (handler)
    handler->reset();

  asio::error_code ignored;
  if (checkTimer)
    checkTimer->cancel(ignored);

  socket.shutdown(asio::ip::tcp::socket::shutdown_both, ignored);
  socket.close(ignored);
}


/*!
 * Чтение тела пакета.
 */
void Connection::Private::readBody(const asio::error_code &err, int bytes)
{
//  qDebug() << "readBody()" << this;
//  qDebug() << "  Transferred:     " << bytes << "bytes";
//  qDebug() << "  RAW Body:        " << QByteArray(bodyBuffer, bodySize).toHex();

  if (!err) {
    if (bytes != bodySize) {
      close();
      return;
    }

    handler->append(opcode, QByteArray(bodyBuffer, bodySize));

//    switch (d->opcode) {
//      case OpcodeGreeting:
//        {
//          quint16 err = opcodeGreeting();
//          if (err) {
//            m_sendQueue.clear();
//            m_sendQueue.enqueue(Packet::create(OpcodeAccessDenied, err));
//            send();
//            handleClose();
//            return;
//          }
//        }
//        m_oldProtocol = true;
//        break;
//    }

//    if (!m_oldProtocol)
//      startPing(WaitPing, schat::PingInterval);

    socket.async_read_some(asio::buffer(headerBuffer, protocol::packet::HeaderSize),
        boost::bind(&Connection::handleReadHeader, q->shared_from_this(), _1, _2));
  }
  else
    close();
}


/*!
 * Чтение заголовка пакета.
 * Размер пакета записывается в \a d->bodySize, тип пакета в \a d->opcode.
 *
 * \param err   Код ошибки.
 * \param bytes Число полученных байт.
 */
void Connection::Private::readHeader(const asio::error_code &err, int bytes)
{
  qDebug() << "readHeader()" << this;
  qDebug() << "  Transferred:     " << bytes << "bytes";
  qDebug() << "  RAW Header:      " << QByteArray(headerBuffer, protocol::packet::HeaderSize).toHex();

  if (!err) {
    using namespace protocol;

    if (bytes != packet::HeaderSize) {
      close();
      return;
    }

    QByteArray header(headerBuffer, packet::HeaderSize);
    QDataStream stream(&header, QIODevice::ReadOnly);
    stream.setVersion(protocol::StreamVersion);
    stream >> bodySize >> opcode;

    if (bodySize > (packet::MaxSize - packet::HeaderSize)) {
      close();
      return;
    }

    if (state() == WaitGreeting) {
      if (!detect()) {
        close();
        return;
      }
      else
        setState(WaitAccess);
    }
    else if (state() != Ready) {
      close();
      return;
    }

//    if (state() != Ready) {
//      if (state() == WaitGreeting && d->opcode == OpcodeGreeting)
//        state(WaitAccess);
//      else if ((state() == WaitGreeting && d->opcode != OpcodeGreeting) || state() == WaitAccess) {
//        handleClose();
//        return;
//      }
//    }

//    qDebug() << "  Full Packet Size:" << d->bodySize + 2 << "bytes";
//    qDebug() << "  Packet Opcode:   " << d->opcode;

    if (bodySize == 0) {
//      if (d->opcode == OpcodePong) {
//        qDebug() << "<<<< PONG OK <<<<" << QDateTime::currentDateTime().toString("[mm:ss]");
//        startPing(WaitPing, schat::PingInterval);
//      }

      socket.async_read_some(asio::buffer(headerBuffer, packet::HeaderSize),
          boost::bind(&Connection::handleReadHeader, q->shared_from_this(), _1, _2));
    }
    else {
      socket.async_read_some(asio::buffer(bodyBuffer, bodySize),
          boost::bind(&Connection::handleReadBody, q->shared_from_this(), _1, _2));
    }
  }
  else
    close();
}


/*!
 * Отправка пакета из очереди.
 * В случае если очередь sendQueue не пуста, то извлекается следующий пакет
 * и записывается в буфер отправки, после этого инициализируется асинхронная
 * запись в сокет.
 *
 * \todo Добавить склейку пакетов.
 */
void Connection::Private::send()
{
//  qDebug() << this << "send()" << QThread::currentThread();

  if (!sendQueue.isEmpty()) {
    QByteArray data = sendQueue.dequeue();
    QDataStream out(&data, QIODevice::ReadOnly);
    out.setVersion(protocol::StreamVersion);
    out.device()->read(sendBuffer, data.size());

//    qDebug() << "  RAW Send:" << QByteArray(m_send, data.size()).toHex();

    socket.async_write_some(asio::buffer(sendBuffer, data.size()),
        boost::bind(&Connection::handleWrite, q->shared_from_this(), _1, _2));
  }
}


/*!
 * Запуск первой асинхронной операции для этого соединения.
 */
void Connection::Private::start()
{
  if (!checkTimer) checkTimer = new asio::deadline_timer(socket.get_io_service());
  checkTimer->expires_from_now(boost::posix_time::seconds(protocol::CheckTimeout));
  checkTimer->async_wait(boost::bind(&Connection::handleCheck, q->shared_from_this(), _1));

  socket.async_read_some(asio::buffer(headerBuffer, protocol::packet::HeaderSize),
      boost::bind(&Connection::handleReadHeader, q->shared_from_this(), _1, _2));
}


bool Connection::Private::detect()
{
  if (opcode == protocol::Greeting) {
    handler = new SimpleClientHandler(q->shared_from_this());
    return true;
  }
  return false;
}


/*!
 * Construct a connection with the given io_service.
 */
Connection::Connection(asio::io_service &ioService)
  : d(new Private(this, ioService))
{
  qDebug() << "Connection()" << this;

//  qRegisterMetaType<UserData>();
//  connect(this, SIGNAL(greeting(const UserData &)), m_daemon, SLOT(greeting(const UserData &)),  Qt::QueuedConnection);
//  connect(this, SIGNAL(leave(const QString &)),     m_daemon, SLOT(localLeave(const QString &)), Qt::QueuedConnection);
}


Connection::~Connection()
{
  qDebug() << "~" << this;
  delete d;
}

/*!
 * Получение сокета ассоциированного с этим соединением.
 */
asio::ip::tcp::socket& Connection::socket()
{
  return d->socket;
}


/*!
 * Закрытие соединения.
 *
 * \note Эта функция потокобезопастна.
 */
void Connection::close()
{
//  m_strand.post(boost::bind(&Connection::handleClose, shared_from_this()));
}


/*!
 * Принудительное закрытие соединения.
 */
void Connection::die()
{
  d->close();
}


/*!
 * Устанавливает состояние соединения в Connection::Ready, что означает
 * успешное прохождение авторизации.
 *
 * \note Эта функция потокобезопастна.
 */
void Connection::ready()
{
  d->setState(Ready);
}


/*!
 * Отправка пакета.
 * Пакет добавляется в очередь Connection::m_sendQueue и если она была пуста
 * немедленно производится начало асинхронной отправки.
 *
 * \param data Полностью сформированный пакет.
 *
 * \note Эта функция потокобезопастна, процесс отправки будет выполнен
 * в контексте этого объекта.
 */
void Connection::send(const QByteArray &data)
{
//  qDebug() << this << "send(const QByteArray &)" << QThread::currentThread();

//  d->mutex.lock();
//  d->sendQueue.enqueue(data);
//  if (d->sendQueue.size() == 1)
//    m_strand.post(boost::bind(&Connection::send, shared_from_this()));

//  d->mutex.unlock();
}


/*!
 * \sa Connection::Private::start().
 */
void Connection::start()
{
  d->start();
}


/*!
 * Обработка пакета \b OpcodeGreeting.
 * Данные извлекаются из пакета для принятия решения о досрочном отказе
 * в доступе, в случае неверных данных.
 *
 * \return \a true в случае успешной предварительной проверки.
 */
quint16 Connection::opcodeGreeting()
{
//  QByteArray body(d->bodyBuffer, d->bodySize - 2);
//  QDataStream stream(&body, QIODevice::ReadOnly);
//  stream.setVersion(StreamVersion);
//
//  quint16 p_version;
//  stream >> p_version;
//  if (p_version < 3)
//    return ErrorOldClientProtocol;
//  else if (p_version > 3)
//    return ErrorOldServerProtocol;
//
//  quint8 p_flag;
//  stream >> p_flag;
//  if (p_flag != FlagNone)
//    return ErrorBadGreetingFlag;
//
//  UserData out;
//
//  stream >> out.gender >> out.nick;
//  out.nick = UserTools::nick(out.nick);
//  if (!UserTools::isValidNick(out.nick))
//    return ErrorBadNickName;
//
//  if (m_daemon->isLocalUser(out.nick))
//    return ErrorNickAlreadyUse;
//
//  stream >> out.fullName >> out.userAgent;
//
//  out.fullName = UserTools::fullName(out.fullName);
//  out.userAgent = UserTools::userAgent(out.userAgent);
//  if (!UserTools::isValidUserAgent(out.userAgent))
//    return ErrorBadUserAgent;
//
//  stream >> out.byeMsg;
//  out.byeMsg = UserTools::byeMsg(out.byeMsg);
//  m_nick = out.nick;
//  out.protocol = 3;
//  out.host = d->socket.remote_endpoint().address().to_string().c_str();
//
//  qDebug() << "           " << m_nick;

  return 0;
}


/*!
 * \sa Connection::Private::check(asio::error_code &err).
 */
void Connection::handleCheck(asio::error_code &err)
{
  d->check(err);
}


/*!
 * \sa Connection::Private::close().
 */
void Connection::handleClose() { d->close(); }


/*!
 * \sa Connection::Private::readBody(const asio::error_code &err, int bytes).
 */
void Connection::handleReadBody(const asio::error_code &err, int bytes)
{
  d->readBody(err, bytes);
}


/*!
 * \sa Connection::Private::readHeader(const asio::error_code &err, int bytes).
 */
void Connection::handleReadHeader(const asio::error_code &err, int bytes)
{
  d->readHeader(err, bytes);
}


/*!
 * Handle completion of a write operation.
 */
void Connection::handleWrite(const asio::error_code &err, int bytes)
{
  Q_UNUSED(bytes)

  if (err)
    d->close();
  else
    d->send();
}


/*!
 * Обработка событий от пинг-таймера.
 * Если текущее состояние равно Connection::WaitPing, то отправляется пинг-пакет
 * и устанавливается состояние ожидания понг-пакета, если понг-таймер
 * не будет сброшен пришедшим понг-пакетом или в случае использования
 * нового протокола любым входящим пакетом, то соединение будет разорвано.
 */
void Connection::ping(asio::error_code &e)
{
  if (!e) {
//    if (d->pingState == WaitPing) {
////      qDebug() << ">>>> PING SEND >>>>" << QDateTime::currentDateTime().toString("[mm:ss]");
//      send(Packet::create(OpcodePing));
//      startPing(WaitPong, schat::PongTimeout);
//    }
//    else {
////      qDebug() << "<<<< PONG FAIL >>>>" << QDateTime::currentDateTime().toString("[mm:ss]");
//      handleClose();
//    }
  }
}


void Connection::startPing(PingState state, int sec)
{
//  if (sec < 1)
//    sec = schat::PingInterval;

  d->pingState = state;
//  d->timer.expires_from_now(boost::posix_time::seconds(sec));
//  d->timer.async_wait(boost::bind(&Connection::ping, shared_from_this(), _1));
}

File: mobile/trunk/src/schat/Connection.java
/* $Id$
 * IMPOMEZIA Simple Chat Mobile
 * Copyright © 2008-2010 IMPOMEZIA <schat@impomezia.com>
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package schat;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.Random;
import java.util.Vector;
import javax.microedition.io.Connector;
import javax.microedition.io.SocketConnection;

/**
 * Класс обеспечивающий подключение к серверу.
 */
public class Connection extends Thread {
  private boolean m_accepted;            // true в случае успешного рукопожатия.
  private boolean m_fatal;               // true если произошла критическая ошибка при подключении.
  private boolean m_state;               // Connection state, true - connected, false - disconnected.
  private DataInputStream m_in;          // input stream.
  private DataOutputStream m_out;        // output stream.
  private int m_port;                    // Порт сервера.
  private long m_rx;                     // Счётчик полученных (receive) байт.
  private long m_tx;                     // Счётчик отправленных (transmit) байт.
  private User m_user;                   // Информация о пользователе.
  private SocketConnection m_connection; // Соединение.
  private String m_host;                 // Адрес сервера.

  private final Object m_transmitLock = new Object();
  private final Vector m_receiveQueue = new Vector(); // Буфер входящих пакетов.


  /**
   * Конструктор класса SimpleClient.
   * @param user
   * @param host
   * @param port
   */
  public Connection(User user, String host, int port) {
    m_user = user;
    m_host = host;
    m_port = port;
    m_state = false;
    m_rx = 0;
    m_tx = 0;
    m_accepted = false;
    m_fatal = false;
  }


  /**
   * Обработка отказа в подключении.
   * @param reason
   */
  private void accessDenied(int reason) {
    System.out.println("accessDenied() " + reason);
    m_state = false;
    m_accepted = false;

    switch (reason) {
      case Protocol.ERROR_OLD_CLIENT_PROTOCOL:
      case Protocol.ERROR_OLD_SERVER_PROTOCOL:
      case Protocol.ERROR_BAD_GREETING_FLAG:
      case Protocol.ERROR_BAD_NICK:
      case Protocol.ERROR_BAD_USER_AGENT:
        enqueuePacket(new RawPacket(reason - 20000, new byte[] {0}));
        m_fatal = true;
        break;

      case Protocol.ERROR_NICK_ALREADY_USE:
        mangleNick();
        break;
    }
  }


  /**
   * Обработка успешного рукопожатия.
   * @param numeric
   */
  private void accessGranted(int numeric) {
    m_accepted = true;
  }


  /**
   * @return Число доступных для чтения пакетов.
   */
  public synchronized int available() {
    return m_receiveQueue.size();
  }


  /**
   * Закрытие соединения.
   */
  public synchronized void close() {
    try { m_in.close(); } catch (Exception e) {}
    m_in = null;

    try { m_out.close(); } catch (Exception e) {}
    m_out = null;

    try { m_connection.close(); } catch (Exception e) {}
    m_connection = null;

    m_state = false;
    m_accepted = false;

    enqueuePacket(new RawPacket(Protocol.VCODE_DISCONNECTED, new byte[] {0}));
  }


  /**
   * Подключение к серверу.
   */
  private boolean connect() {
    m_state = false;
    m_fatal = false;

    enqueuePacket(new RawPacket(Protocol.VCODE_CONNECTING, new byte[] {0}));

    try {
      m_connection = (SocketConnection) Connector.open("socket://" + m_host + ":" + m_port, Connector.READ_WRITE, true);
      m_in = m_connection.openDataInputStream();
      m_out = m_connection.openDataOutputStream();

      m_state = true;

    } catch (Exception e) {
      //#ifdef SCHAT_DEBUG
      System.out.println("Error trying to connect to server, aborting...");
      System.out.println("Exception: " + e.getMessage());
      e.printStackTrace();
      //#endif
    }
    finally {
      if (!m_state) close();
    }

    return m_state;
  }


  /**
   * Запускает подключение к серверу.
   * Попытка соединения осуществляется каждые 6 секунд.
   */
  private void connectLoop() {
    do {
      if (m_fatal)
        break;

      if (!connect()) {
        try {
          sleep(6 * 1000);
        } catch (Exception e) {
        }
      }
    } while (!m_state);
  }


  /**
   * Добавление пакета в очередь.
   * @param packet
   */
  private void enqueuePacket(RawPacket packet) {
    synchronized (m_receiveQueue) {
      m_receiveQueue.addElement(packet);
    }

    synchronized (SimpleClient.waitObj()) {
      SimpleClient.waitObj().notify();
    }
  }


  /**
   * Создаёт новый ник для устранения коллизии.
   */
  private void mangleNick() {
    int max = 99;
    String nick = Simple.storage().nick;
    if (nick.length() == User.MAX_NICK_LENGTH)
      nick = nick.substring(0, User.MAX_NICK_LENGTH - 2);
    else if (nick.length() == User.MAX_NICK_LENGTH - 1)
      max = 9;

    setNick(nick + new Random().nextInt(max));
  }


  /**
   * @return Возвращает ник.
   */
  public synchronized String nick() {
    return m_user.nick();
  }


  /**
   * @return Возвращает следующий пакет или null если больше нет пакетов.
   */
  public RawPacket packet() {
    RawPacket packet;

    synchronized (m_receiveQueue) {
      if (m_receiveQueue.size() == 0) {
        return null;
      }
      packet = (RawPacket) this.m_receiveQueue.elementAt(0);
      this.m_receiveQueue.removeElementAt(0);
    }

    return packet;
  }


  /**
   * Цикл чтения данных.
   */
  private void read() {
    int size;
    int opcode;
    m_accepted = false;

    synchronized (m_receiveQueue) {
      m_receiveQueue.removeAllElements();
    }

    try {
      while (m_state) {
        size = m_in.readUnsignedShort();
        opcode = m_in.readUnsignedShort();
        m_rx += size + 2;

        if (m_accepted) {
          if (size > 2) {
            byte[] body = new byte[size - 2];
            m_in.readFully(body);
            enqueuePacket(new RawPacket(opcode, body));
          }
          else if (size == 2) {
            switch (opcode) {
              case Protocol.OPCODE_PING:
                sendPacket(new RawPacket(Protocol.OPCODE_PONG, null));
                break;

              case Protocol.OPCODE_SYNC_USERS_END:
                enqueuePacket(new RawPacket(opcode, new byte[] {0}));
                break;
            }
          }
        }
        else {
          if (opcode == Protocol.OPCODE_ACCESS_GRANDED)
            accessGranted(m_in.readUnsignedShort());
          else if (opcode == Protocol.OPCODE_ACCESS_DENIED)
            accessDenied(m_in.readUnsignedShort());
          else
            break;
        }
      }
    } catch (Exception e) {
//      e.printStackTrace();
    }

    close();
  }


  /**
   * Основной цикл потока.
   */
  public void run() {
    System.out.println("Connection runned");

    do {
      connectLoop();
      if (m_state) {
        sendPacket(Protocol.greetingPacket(m_user));
        read();
      }

      if (!m_fatal) {
        try {
          sleep(1000);
        } catch (Exception e) {
        }
      }
    } while (!m_fatal);

    System.out.println("Connection thread loop end");
  }


  /**
   * @return Число полученных байт.
   */
  public synchronized long rx() {
    return m_rx;
  }


  /**
   * Отправка сырого пакета.
   * @param packet
   * @return false если при отправке произошла ошибка.
   */
  public boolean sendPacket(RawPacket packet) {
    if (packet == null || m_out == null)
      return false;

    synchronized (m_transmitLock) {
      try {
        byte[] data = packet.toByteArray();
        m_out.write(data);
        m_out.flush();
        m_tx += data.length;

      } catch (IOException e) {
        close();
        return false;
      }
    }

    return true;
  }


  /**
   * Установка критической ошибки.
   * @param fatal
   */
  public synchronized void setFatal(boolean fatal) {
    m_fatal = fatal;
  }


  /**
   * Устанавливает ник.
   * @param nick
   */
  public synchronized void setNick(String nick) {
    m_user.setNick(nick);

    if (m_accepted)
      sendPacket(Protocol.newProfilePacket(m_user));
  }


  /**
   * Connection state.
   * @return true - connected, false - disconnected.
   */
  public synchronized boolean state() {
    return m_state;
  }


  /**
   * @return Число отправленных байт.
   */
  public synchronized long tx() {
    return m_tx;
  }
}
